pkg/autoscaling/calculation.go

```diff
@@ -222,20 +221,44 @@ func getScaledTiKVGroups(informer core.StoreSetInformer, healthyInstances []inst
 			return nil
 		}
 		v := store.GetLabelValue(groupLabelKey)
-		if len(v) > len(autoScalingGroupLabelKeyPrefix) &&
-			v[:len(autoScalingGroupLabelKeyPrefix)] == autoScalingGroupLabelKeyPrefix {
-			if stores, ok := planMap[v]; ok {
-				stores[instance.id] = struct{}{}
-			} else {
-				planMap[v] = map[uint64]struct{}{
-					instance.id: {},
-				}
+		buildPlanMap(planMap, v, instance.address)
+	}
+	return buildPlans(planMap, TiKV)
+}
+
+func getScaledTiDBGroups(informer core.TiDBInformer, healthyInstances []instance) []*Plan {
+	planMap := make(map[string]map[string]struct{}, len(healthyInstances))
+	for _, instance := range healthyInstances {
+		tidb := informer.GetTiDB(instance.address)
+		if tidb == nil {
+			log.Warn("inconsistency between health instances and tidb status, exit auto-scaling calculation",
+				zap.String("tidb-address", instance.address))
+			return nil
+		}
+		v := tidb.GetLabelValue(groupLabelKey)
+		buildPlanMap(planMap, v, instance.address)
+	}
+	return buildPlans(planMap, TiDB)
+}
+
+func buildPlanMap(planMap map[string]map[string]struct{}, groupName, address string) {
+	if len(groupName) > len(autoScalingGroupLabelKeyPrefix) &&
+		groupName[:len(autoScalingGroupLabelKeyPrefix)] == autoScalingGroupLabelKeyPrefix {
+		if tidb, ok := planMap[groupName]; ok {
+			tidb[address] = struct{}{}
```

Comment:
I think `tidb` is a little bit misleading here

Reply:
nice catch. updated.